### ISSUE TITLE
ISPAC File Path Access Denied

### ISSUE THREAD

**Issue by BankGuy** [ToxicGlobe/VSTS-SSIS-Extension#19]
Hello. I am encountering an error when using the Deploy step in a Release Definition in TFS 2015. SSDT is installed on the build box. The error in the log is as follows:

> ##[error]Task_InternalError Exception calling "ReadAllBytes" with "1" argument(s): "Access to the path 'C:\TFSAgents\agent1\_work\...\DEV' is denied."

I looked at the underlying PowerShell script, and it looks like the exception is being thrown by the following line:

> [byte[]] $projectFile = [System.IO.File]::ReadAllBytes($ProjectFilePath)

From what I can tell, the account that the TFS Agents run as has Full Control permissions, so I do not understand why I am getting an access denied error.

Please help.
Thanks!

**Comment by schwaizi** [comment]
Hi BankGuy, first thanks for your feedback.

From the error message i assume your agent/script can not read the project file.

I would check if the file exist. Eventually log on uaing the agent service account and try it directly.

I guess it is security configuration issue on the box, not a bug with the script.

Another thing, having TFS Agent on system drive is not the best option. It may fill up your drive with temp files. And it may be the reason for the access denied as system drive may be controlled by UAC.

Cheers Peter

**Comment by BankGuy** [comment]
Yes, we are painfully aware that using the system drive is a bad idea. We are in the process of figuring out how to move the Agent work area to another location. Having said that, I did verify that UAC is off.

I just added a Powershell step before the SSIS Deploy step to do two things: 1) run whoami to verify the account I'm using is the one I was expecting, and 2) run CACLS to verify that the account has Full Control. This step completed successfully, which also proves that the ISPAC file in question not only exists, but is accessible to the agent service account.

(Someone here suggested that I verify that another process hasn't locked the file. It is worth noting that the filename of the ISPAC file includes the TFS build number, so this file gets recreated from scratch every build. Moreover, the build artifacts are published to the server at the end of the build and then brought back down to another directory in the agent work area for deployment. I would expect that any if not all of these factors would pretty-much guarantee that the ISPAC file is fresh and not locked.)

Last, I added another Powershell step, again right before the SSIS Deploy step, this time to execute the ReadAllBytes method in isolation.  Here is the command:

> [byte[]] $projectFile = [System.IO.File]::ReadAllBytes('C:\TFSAgents\agent1\_work\...\$(Build.DefinitionName)\$(Release.EnvironmentName)\$(Build.BuildNumber).ispac')

The log showed no output (so, no errors - just one line saying that it was about to execute a PowerShell script), which I assume means that the file was read into the byte array correctly.

So, now the question is, why does the command work in isolation but fail within the SSIS Deploy step?

**Comment by schwaizi** [comment]
Hi BankGuy,

prety nice analysis! Now is getting more interesting!

I took a look in the script (you can see it here on GitHub) and there is simple ReadAllBytes command. So wondering what the issue is.
One point is there are many installation of this extension and noone else have the same issue. I know Hosted 2017 agent is not working but you have it on-premise so it’s not your case.

I also can not reproduce the issue and i have it on C drive on my laptop.

I really believe it is some security settings on your box, but i do not know what may cause this issue.

Can you try another box (your laptop for example) or hosted agent?

Regarding the agent _work folder re-location. You can run config.cmd and re-enter the values where one of those is _work folder location.

**Comment by BankGuy** [comment]
Thanks for the quick reply. I am out of the office tomorrow, so it will be next week before I can look at it again. Until then, have a good weekend!

**Comment by BankGuy** [comment]
Sorry for the delay. The good news is that the deployment works now! We incorrectly assumed that the input parameter labeled 'Path to .ispac file' literally meant the _path to the directory containing the ISPAC file_ and was not inclusive of the filename. Adding the file name to the end of the path solved the problem. Adding clarification via a little information button/tool tip to the right of the field may prevent others from making the same mistake we did.

Also, I looked back at the script to see if there was any input validation to catch this. Immediately after the SQL Connection String is printed to the screen, there _is_ a test to verify that $projectFilePath _exists_, but there is _no_ check to verify that it is a _file_. The directory does exist, of course, but apparently passing a directory name to the ReadAllBytes method returns the very misleading error about the access being denied.

We ran into another issue downstream from that, but I will raise it as a separate issue if it hasn't already been submitted by someone else.

Thanks again for the help!

**Comment by BankGuy** [comment]
Sorry for the delay. The good news is that the deployment works now! We incorrectly assumed that the input parameter labeled 'Path to .ispac file' literally meant the _path to the directory containing the ISPAC file_ and was not inclusive of the filename. Adding the file name to the end of the path solved the problem. Adding clarification via a little information button/tool tip to the right of the field may prevent others from making the same mistake we did.

Also, I looked back at the script to see if there was any input validation to catch this. Immediately after the SQL Connection String is printed to the screen, there _is_ a test to verify that $projectFilePath _exists_, but there is _no_ check to verify that it is a _file_. The directory does exist, of course, but apparently passing a directory name to the ReadAllBytes method returns the very misleading error about the access being denied.

We ran into another issue downstream from that, but I will raise it as a separate issue if it hasn't already been submitted by someone else.

Thanks again for the help!


**********************************************************************
This email and any files transmitted with it are confidential and may contain protected or privileged material. If you are not the intended recipient be advised that you have received this email in error and that any use, dissemination, forwarding, printing or copying of this email is strictly prohibited.

**Comment by markosaric** [comment]
What permissions do we need to be granted the file path used in TFS release deployment pipeline for this task to work? I've tried to grant permissions for everyone, agent account user, etc... but I've got the following message:

##[error]Task_InternalError Cannot access project file path: C:\Deployment\SSIS\Demo.ispac

The project path is on the database server and the file along with the folder exists.
